types/index.d.ts

```diff
@@ -0,0 +1,496 @@
+
+    /** Whether or not Mirage should log all requests/response cycles. */
+    logging: boolean;
+
+    /** A default number of ms to artificially delay responses for all routes. */
+    timing: number;
+
+    /** A default prefix applied to all subsequent route definitions. */
+    namespace: string;
+
+    /** Sets a string to prefix all route handler URLs with. */
+    urlPrefix: string;
+
+    /** Actual Pretender instance */
+    pretender: PretenderServer;
+
+    /** Creates multiple models of the given type. */
+    createList<
+      K extends keyof Registry,
+      Init extends Registry[K],
+      Data extends Partial<Init>
+    >(modelName: K, count: number, data?: Data): Array<Init & Data>;
+
+    /** Handle a GET request to the given path. */
+    get(
+      path: string,
+      handler?: RouteHandler<Registry>,
+      options?: HandlerOptions
+    ): void;
+
+    /** Handle a POST request to the given path. */
+    post(
+      path: string,
+      handler?: RouteHandler<Registry>,
+      options?: HandlerOptions
+    ): void;
+
+    /** Handle a PUT request to the given path. */
+    put(
+      path: string,
+      handler?: RouteHandler<Registry>,
+      options?: HandlerOptions
+    ): void;
+
+    /** Handle a PATCH request to the given path. */
+    patch(
+      path: string,
+      handler?: RouteHandler<Registry>,
+      options?: HandlerOptions
+    ): void;
+
+    /** Handle an OPTIONS request to the given path. */
+    options(
+      path: string,
+      handler?: RouteHandler<Registry>,
+      options?: HandlerOptions
+    ): void;
+
+    /** Handle a DELETE request to the given path. */
+    del(
+      path: string,
+      handler?: RouteHandler<Registry>,
+      options?: HandlerOptions
+    ): void;
+
+    delete(
+      path: string,
+      handler?: RouteHandler<Registry>,
+      options?: HandlerOptions
+    ): void;
+
+    /** Pass through one or more URLs to make real requests. */
+    passthrough(urls?: ((request: Request) => any) | string | string[]): void;
+
+    /** Load all available fixture data matching the given name(s). */
+    loadFixtures(...names: string[]): void;
+
+    seeds(server: Server): void;
+
+    routes(): void;
+
+    /** Shutdown the server and stop intercepting network requests. */
+    shutdown(): void;
+  }
+}
+
+declare module "miragejs/db" {
+  import DbCollection from "miragejs/db-collection";
+  import IdentityManager from "miragejs/identity-manager";
+
+  /** The in-memory database containing all currently active data keyed by collection name. */
+  export default class Db {
+    [key: string]: DbCollection | any;
+
+    constructor(initialData: [], identityManagers?: IdentityManager[]);
+
+    createCollection(name: string, initialData?: any[]): void;
+    dump(): void;
+    emptyData(): void;
+    loadData(data: any): void;
+  }
+}
+
+declare module "miragejs/db-collection" {
+  import IdentityManager from "miragejs/identity-manager";
+  export default class DbCollection {
```

Comment:
We may want to circle back and make this generic so that it can wrap over a specific type—that can come in a follow-on pass, but note that doing so *will* technically constitute a breaking change to the types, so we may want to do it here and now.

```ts
export default class DbCollection<T> {
  constructor(
    name: string,
    initialData: T[],
    identityManager?: IdentityManager
  );

  all(): T[];

  // etc.
```

Note that this will likely also involve plumbing that info back through the rest of the system.

Given that would be a breaking change, we may want to land this particular PR, *then* do that work, but make sure to land it *before* cutting a Mirage release with the type defs published.

Reply:
Good suggestion, probably it is worth to open an issue ticket to keep as a reminder.

Reply:
Opened an issue [here](https://github.com/miragejs/miragejs/issues/380)!

Reply:
Thanks @chriskrycho Great. Let's merge the actual PR and after that we can migrate to generic DbCollection. Cool stuff. :)